### PR TITLE
US156 - As a developer, I want a DiceManager to handle player dice rolls so that movement distance is determined correctly.

### PR DESCRIPTION
**Pull request by tsaldrid**
Task 157 - Implement DiceManagerClass
Task 158 - Integrate with PlayerTurn() flow in GameManager
Task 159 - Create DiceRolledEvent ScriptableObject Instance
Task 160 - Create unit tests for roll range and event triggers

This was rebased with US145 to use the new logger functionality.

### CODE REVIEW

**Comment by kescob0825** [comment]
# Code Review Checklist

**Completed for**: US #156 and Tasks #157 to #160  
**Code Review Completed By**: Karl Escobar

Use this checklist to review new code changes before merging.

---

## Implementation
- [x] Does this code change accomplish what it is supposed to do?
- [x] Can the solution be simplified?
- [x] This change does **not** add unwanted compile-time or run-time dependencies.
- [x] The code is at the right abstraction level.
- [x] The code is modular enough.
- [x] The solution is adequate for maintainability, readability, performance, and security.
- [x] Similar functionality does **not** already exist in the codebase (or reuse is not appropriate).
- [x] No additional best practices or design patterns would substantially improve this code.

---

## Logic Errors and Bugs
- [x] No use cases where the code behaves incorrectly.
- [x] No inputs or external events that can break the code.
- [x] Code agrees with specifications.
- [x] No off-by-one errors.
- [x] No magic numbers (or they are well-named constants).
  - Justification: The magic numbers in the codebase are temporary placeholders. The developer added comments for justification.
- [x] Proper use of global/large-scope variables (minimized and justified).

---

## Error Handling and Logging
- [x] Error handling is implemented correctly.
- [x] Error messages are user-friendly (where applicable).
- [x] Sufficient, meaningful log events exist for debugging (no sensitive data in logs).

---

## Dependencies
- [x] Documentation/configuration/README updates were made as needed.
- [x] No negative impact on other parts of the system or backward compatibility.

---

## Security and Data Privacy
- [x] No new security vulnerabilities introduced.
- [x] All user input is validated, sanitized, and escaped (XSS/SQL injection prevented).
- [x] Data from external APIs/libraries is handled safely and checked as needed.
---

## Performance
- [x] No significant performance regressions.
- [x] No obvious opportunities for significant performance gains left unaddressed.

---

## Ethics and Morality
- [x] Appropriate measures exist to prevent or report harassment/abuse (if applicable).

---

## Testing and Testability
- [x] Code is testable.
- [x] Automated tests have been added or updated.
- [x] Existing tests sufficiently cover the change (unit/integration/system).
- [x] Edge cases and boundary inputs are tested.

---

## Readability
- [x] Code is easy to understand.
- [x] Methods/functions are appropriately small and focused.
- [x] Names (functions/classes/variables) are clear and descriptive.
- [x] Code is in the correct file/folder/package.
- [x] Control flow is intuitive; data flow is clear.
- [x] Comments are clear, relevant, and up to date.
- [x] No commented-out or redundant code remains.
- [x] No overly dense one-liners; no overly long methods.

File: Assets/Scripts/Managers/DiceManager.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        dieOne = Random.Range(MIN, MAX);
+        dieTwo = Random.Range(MIN, MAX);
+        totalRoll = dieOne + dieTwo;
+
+        DiceRolledEvent diceRolledEvent = new DiceRolledEvent(dieOne, dieTwo, totalRoll);
+
+        //Tests that the diceRolledChannel isn't null, and raises an event
+        if (diceRolledChannel != null)
+        {
+            diceRolledChannel?.RaiseEvent(diceRolledEvent);
```
Comment (hdathert) [inline Assets/Scripts/Managers/DiceManager.cs]: The null check here is redundant with the `?` syntax sugar. Given that we want an exception thrown if it fails, the `?` isn't really needed at all.

File: Assets/Tests/EditMode/GameManagerTests/GameManagerTestBase.cs [hunk 2]
```diff
@@ -21,11 +22,23 @@ public virtual void SetUp()
...
+
+            InitializeTestLogger();
+
         }
 
         [TearDown]
         public virtual void TearDown()
         {
+            //unsubscribe to event channel.
+            gameManager.diceRolledChannel.Unsubscribe(gameManager.DiceRolled);
```
Comment (hdathert) [inline Assets/Tests/EditMode/GameManagerTests/GameManagerTestBase.cs]: This isn't needed, since `DestroyTestObjects` destroys the event channels as well.

File: ProjectSettings/Packages/com.unity.testtools.codecoverage/Settings.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
             {
                 "type": "System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089",
                 "key": "IncludeAssemblies",
-                "value": "{\"m_Value\":\"Assembly-CSharp,Assembly-CSharp-Editor,PsycheOpoly.Runtime\"}"
+                "value": "{\"m_Value\":\"Assembly-CSharp,Assembly-CSharp-Editor,PsycheOpoly.Runtime,Tests\"}"
```
Comment (hdathert) [inline ProjectSettings/Packages/com.unity.testtools.codecoverage/Settings.json]: We don't want to have Tests here, because then the Code Coverage tool will assess our test files for code coverage as well.

**Review by hdathert — CHANGES_REQUESTED** [review]
There are a couple of small changes needed to maintain style/pattern adherence. There's also a merge conflict happening on `dev` for some reason- that needs a look over.

**Review by hdathert — APPROVED** [review]
# Code Review Checklist

**Completed for**: US #156 and Tasks #157 to #160 
**Code Review Completed By**: @hdathert 

Use this checklist to review new code changes before merging.

---

## Implementation
- [x] Does this code change accomplish what it is supposed to do?
- [x] Can the solution be simplified?
- [x] This change does **not** add unwanted compile-time or run-time dependencies.
- [x] The code is at the right abstraction level.
- [x] The code is modular enough.
- [x] The solution is adequate for maintainability, readability, performance, and security.
- [x] Similar functionality does **not** already exist in the codebase (or reuse is not appropriate).
- [x] No additional best practices or design patterns would substantially improve this code.

---

## Logic Errors and Bugs
- [x] No use cases where the code behaves incorrectly.
- [x] No inputs or external events that can break the code.
- [x] Code agrees with specifications.
- [x] No off-by-one errors.
- [x] No magic numbers (or they are well-named constants).
- [x] Proper use of global/large-scope variables (minimized and justified).

---

## Error Handling and Logging
- [x] Error handling is implemented correctly.
- [x] Error messages are user-friendly (where applicable).
- [x] Sufficient, meaningful log events exist for debugging (no sensitive data in logs).

---

## Dependencies
- [x] Documentation/configuration/README updates were made as needed.
- [x] No negative impact on other parts of the system or backward compatibility.

---

## Security and Data Privacy
- [x] No new security vulnerabilities introduced.
- [x] All user input is validated, sanitized, and escaped (XSS/SQL injection prevented).
- [x] Data from external APIs/libraries is handled safely and checked as needed.

---

## Performance
- [x] No significant performance regressions.
- [x] No obvious opportunities for significant performance gains left unaddressed.

---

## Ethics and Morality
- [x] Appropriate measures exist to prevent or report harassment/abuse (if applicable).

---

## Testing and Testability
- [x] Code is testable.
- [x] Automated tests have been added or updated.
- [x] Existing tests sufficiently cover the change (unit/integration/system).
- [x] Edge cases and boundary inputs are tested.

---

## Readability
- [x] Code is easy to understand.
- [x] Methods/functions are appropriately small and focused.
- [x] Names (functions/classes/variables) are clear and descriptive.
- [x] Code is in the correct file/folder/package.
- [x] Control flow is intuitive; data flow is clear.
- [x] Comments are clear, relevant, and up to date.
- [x] No commented-out or redundant code remains.
- [x] No overly dense one-liners; no overly long methods.

---

**Use this before submitting a pull request.**  
It helps ensure code quality, maintainability, and compliance with best practices.

*Based on awesome code review guides and MIT’s code review recommendations.*
